### PR TITLE
feat: add crate for pub modifier

### LINKED ISSUE #3168: Add pub(crate) visibility
### Problem

We support the `pub` attribute when defining a function, so that the function is visible everywhere externally and internally.

We could extend it with the `pub(crate)` notation which would allow the function to be used inside the crate, but not in an external crate.

### Happy Case

`pub(crate)` makes a function visible within the current crate.

### Alternatives Considered

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Problem\*

Functions are either public or private. With this PR they can be restricted to their crate.

Resolves #3168 

## Documentation

Documentation update is included in this PR.

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: docs/docs/language_concepts/01_functions.md [hunk 1]
```diff
@@ -20,6 +20,12 @@ By default, functions are visible only within the package they are defined. To m
 pub fn foo() {}
 ```
 
+You can however restrict the visibility of the function to its own crate, by specifying `pub(crate)`:
```
Comment (jfecher) [inline docs/docs/language_concepts/01_functions.md:23]:
```suggestion
You can also restrict the visibility of the function to only the crate it was defined in, by specifying `pub(crate)`:
```

File: compiler/noirc_frontend/src/parser/parser.rs [hunk 2]
```diff
@@ -198,19 +204,34 @@ fn function_definition(allow_self: bool) -> impl NoirParser<NoirFunction> {
...
+                open.is_some(),
+                internal.is_some(),
+                public.is_some(),
+            )
         })
 }
+fn is_pub_crate() -> impl NoirParser<bool> {
+    (keyword(Keyword::Pub)
+        .then_ignore(just(Token::LeftParen))
+        .then(keyword(Keyword::Crate))
```
Comment (jfecher) [inline compiler/noirc_frontend/src/parser/parser.rs:230]:
```suggestion
        .then_ignore(keyword(Keyword::Crate))
```

**Review by jfecher — APPROVED** [review]
LGTM

**Comment by Savio-Sou** [comment]
cc @signorecello as this involved doc addition

**Pull request merged**